File: backend/Concord/Concord.Application/Models/MessageResponse.cs
using System.Text.Json.Serialization;

namespace Concord.Application.Models;

public class MessageResponse
{
    [JsonPropertyName("Id")]
    public Guid Id { get; set; }

    [JsonPropertyName("ChannelId")]
    public Guid ChannelId { get; set; }

    [JsonPropertyName("Sender")]
    public PublicProfileResponse Sender { get; set; } = null!;

    [JsonPropertyName("Content")]
    public string? Content { get; set; }

    [JsonPropertyName("Created")]
    public DateTime Created { get; set; }

    [JsonPropertyName("EditedAtUtc")]
    public DateTime? EditedAtUtc { get; set; }

    [JsonPropertyName("AttachmentUrl")]
    public string? AttachmentUrl { get; set; }

    [JsonPropertyName("ReplyToMessageId")]
    public Guid? ReplyToMessageId { get; set; }

    [JsonPropertyName("Reactions")]
    public List<ReactionSummaryResponse> Reactions { get; set; } = [];

    [JsonPropertyName("PinnedAt")]
    public DateTime? PinnedAt { get; set; }

    [JsonPropertyName("PinnedByUserId")]
    public Guid? PinnedByUserId { get; set; }

    [JsonPropertyName("ForwardedFromSenderId")]
    public Guid? ForwardedFromSenderId { get; set; }

    [JsonPropertyName("ForwardedFromCreatedAt")]
    public DateTime? ForwardedFromCreatedAt { get; set; }

    [JsonPropertyName("MentionedUserIds")]
    public List<Guid> MentionedUserIds { get; set; } = [];

    [JsonPropertyName("MentionsEveryone")]
    public bool MentionsEveryone { get; set; }
}

File: mobile/concord/lib/api/models/message_response.dart
import 'json_utils.dart';
import 'message_like.dart';
import 'public_profile_response.dart';
import 'reaction_summary_response.dart';

class MessageResponse implements MessageLike {
  const MessageResponse({
    required this.id,
    required this.channelId,
    required this.sender,
    required this.content,
    required this.created,
    required this.editedAtUtc,
    required this.attachmentUrl,
    required this.replyToMessageId,
    required this.reactions,
    required this.pinnedAt,
    required this.pinnedByUserId,
    required this.forwardedFromSenderId,
    required this.forwardedFromCreatedAt,
    required this.mentionedUserIds,
  });

  factory MessageResponse.fromJson(Map<String, dynamic> json) {
    final rawReactions = json.field('Reactions') as List<dynamic>? ?? const [];
    return MessageResponse(
      id: json.field('Id') as String,
      channelId: json.field('ChannelId') as String,
      sender: PublicProfileResponse.fromJson(json.field('Sender') as Map<String, dynamic>),
      content: json.field('Content') as String?,
      created: parseDateTime(json.field('Created')),
      editedAtUtc: parseNullableDateTime(json.field('EditedAtUtc')),
      attachmentUrl: json.field('AttachmentUrl') as String?,
      replyToMessageId: json.field('ReplyToMessageId') as String?,
      reactions: rawReactions
          .map((e) => ReactionSummaryResponse.fromJson(e as Map<String, dynamic>))
          .toList(),
      pinnedAt: parseNullableDateTime(json.field('PinnedAt')),
      pinnedByUserId: json.field('PinnedByUserId') as String?,
      forwardedFromSenderId: json.field('ForwardedFromSenderId') as String?,
      forwardedFromCreatedAt: parseNullableDateTime(json.field('ForwardedFromCreatedAt')),
      mentionedUserIds: parseStringList(json.field('MentionedUserIds')),
    );
  }

  @override
  final String id;
  final String channelId;
  @override
  final PublicProfileResponse sender;

  @override
  final String? content;
  @override
  final DateTime created;
  @override
  final DateTime? editedAtUtc;
  @override
  final String? attachmentUrl;
  @override
  final String? replyToMessageId;
  @override
  final List<ReactionSummaryResponse> reactions;
  @override
  final DateTime? pinnedAt;
  @override
  final String? pinnedByUserId;
  @override
  final String? forwardedFromSenderId;
  @override
  final DateTime? forwardedFromCreatedAt;
  @override
  final List<String> mentionedUserIds;
}
